e2e-polybft/framework/test-cluster.go

```diff
@@ -554,3 +591,185 @@ func (c *TestCluster) InitSecrets(prefix string, count int) ([]types.Address, er
 
 	return result, nil
 }
+
+var (
+	defaultGasPrice = uint64(1879048192) // 0x70000000
+	defaultGasLimit = uint64(5242880)    // 0x500000
+)
+
+func (t *TestCluster) ExistsCode(tt *testing.T, addr ethgo.Address) bool {
+	client, err := jsonrpc.NewClient(t.Servers[0].JSONRPCAddr())
+	require.NoError(tt, err)
+
+	code, err := client.Eth().GetCode(addr, ethgo.Latest)
+	if err != nil {
+		return false
+	}
+	if code == "0x" {
+		return false
+	}
+
+	return true
+}
+
+func (t *TestCluster) Call(tt *testing.T, to types.Address, method *abi.Method, args ...interface{}) map[string]interface{} {
+	client, err := jsonrpc.NewClient(t.Servers[0].JSONRPCAddr())
+	require.NoError(tt, err)
+
+	input, err := method.Encode(args)
+	require.NoError(tt, err)
+
+	toAddr := ethgo.Address(to)
+
+	msg := &ethgo.CallMsg{
+		To:   &toAddr,
+		Data: input,
+	}
+	resp, err := client.Eth().Call(msg, ethgo.Latest)
+	require.NoError(tt, err)
+
+	data, err := hex.DecodeString(resp[2:])
+	require.NoError(tt, err)
+
+	output, err := method.Decode(data)
+	require.NoError(tt, err)
+
+	return output
+}
+
+type TestCall struct {
+}
+
+func (t *TestCluster) Deploy(tt *testing.T, sender ethgo.Key, bytecode []byte) *TestTxn {
+	return t.SendTxn(tt, sender, &ethgo.Transaction{Input: bytecode})
+}
+
+func (t *TestCluster) Transfer(tt *testing.T, sender ethgo.Key, target types.Address, value *big.Int) *TestTxn {
+	targetAddr := ethgo.Address(target)
+	return t.SendTxn(tt, sender, &ethgo.Transaction{To: &targetAddr, Value: value})
+}
+
+func (t *TestCluster) MethodTxn(tt *testing.T, sender ethgo.Key, target types.Address, input []byte) *TestTxn {
+	targetAddr := ethgo.Address(target)
+	return t.SendTxn(tt, sender, &ethgo.Transaction{To: &targetAddr, Input: input})
+}
+
+// SendTxn sends a transaction
+func (t *TestCluster) SendTxn(tt *testing.T, sender ethgo.Key, txn *ethgo.Transaction) *TestTxn {
```

Comment:
This is a wrapper around the execution of transactions in the `e2e` framework. I tried to use the relayer but I found it more helpful to have a custom thing on `TestCluster` to which we can attach specific methods and wrappers for testing.